src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs

```diff
@@ -770,27 +770,23 @@ public static void Reverse(ref char buf, nuint length)
                     tempLast = Avx2.Permute2x128(tempLast, tempLast, 0b00_01);
 
                     // Store the reversed vectors
-                    tempLast.StoreUnsafe(ref bufByte, firstOffset);
-                    tempFirst.StoreUnsafe(ref bufByte, lastOffset);
-                }
-                bufByte = ref Unsafe.Add(ref bufByte, numIters * numElements);
-                length -= numIters * (nuint)Vector256<short>.Count * 2;
-                // Store any remaining values one-by-one
-                buf = ref Unsafe.As<byte, char>(ref bufByte);
+                    tempLast.StoreUnsafe(ref Unsafe.As<char, byte>(ref first));
+                    tempFirst.StoreUnsafe(ref Unsafe.As<char, byte>(ref last));
+
+                    first = ref Unsafe.Add(ref first, Vector256<ushort>.Count);
+                    last = ref Unsafe.Subtract(ref last, Vector256<ushort>.Count);
+                } while (!Unsafe.IsAddressGreaterThan(ref first, ref last));
+
+                // shift for division is fine here since we don't care about what any negative number end up being
+                remainder = Unsafe.ByteOffset(ref first, ref Unsafe.Add(ref last, Vector256<ushort>.Count)) >> 1;
             }
-            else if (Vector128.IsHardwareAccelerated && (nuint)Vector128<short>.Count * 2 <= length)
+            else if (Vector128.IsHardwareAccelerated && length >= (nuint)Vector128<ushort>.Count)
             {
-                ref short bufShort = ref Unsafe.As<char, short>(ref buf);
-                nuint numElements = (nuint)Vector128<short>.Count;
-                nuint numIters = (length / numElements) / 2;
-                for (nuint i = 0; i < numIters; i++)
+                last = ref Unsafe.Subtract(ref last, Vector128<ushort>.Count);
+                do
                 {
-                    nuint firstOffset = i * numElements;
-                    nuint lastOffset = length - ((1 + i) * numElements);
-
-                    // Load in values from beginning and end of the array.
-                    Vector128<short> tempFirst = Vector128.LoadUnsafe(ref bufShort, firstOffset);
-                    Vector128<short> tempLast = Vector128.LoadUnsafe(ref bufShort, lastOffset);
+                    Vector128<ushort> tempFirst = Vector128.LoadUnsafe(ref Unsafe.As<char, ushort>(ref first));
```

Comment:
As I'm sure u know - the casts themselves doesn't have any impact on the generated code, so we're dealing with just a design choice. I think that it makes more sense, and is more consistent, to keep the `ref char` input since we are in the SpanHelper.Char context, and the caller would naturally want to pass a char